dds/DCPS/DataReaderImpl.cpp

```diff
@@ -3570,6 +3540,185 @@ const DataReaderImpl::HandleSet& DataReaderImpl::lookup_matching_instances(CORBA
+
+          // We need to update the last total count value to our current total
+          // so that the next time we will calculate the correct total_count_change;
+          last_deadline_missed_total_count_ = requested_deadline_missed_status_.total_count;
+        }
+
+        notify_status_condition();
+      }
+    }
+
+    // This next part is without status_lock_ held to avoid reactor deadlock.
+    if (timer_called) {
+      instance->deadline_ = MonotonicTimePoint::zero_value;
+      schedule_deadline(instance, timer_called);
+    } else {
+      cancel_deadline(instance);
+      schedule_deadline(instance, timer_called);
+    }
+  }
+}
+
+void DataReaderImpl::cancel_all_deadlines()
+{
+  ACE_GUARD(ACE_Recursive_Thread_Mutex, guard, sample_lock_);
+  deadline_queue_.clear();
+  deadline_task_->cancel();
+}
+
+void DataReaderImpl::reset_deadline_period(const TimeDuration& deadline_period)
+{
+  if (deadline_period_ != deadline_period) {
+    deadline_period_ = deadline_period;
+
+    if (deadline_queue_enabled_) {
+      ACE_GUARD(ACE_Recursive_Thread_Mutex, instance_guard, this->instances_lock_);
+      const MonotonicTimePoint now = MonotonicTimePoint::now();
+      for (SubscriptionInstanceMapType::iterator iter = this->instances_.begin();
+           iter != this->instances_.end();
+           ++iter) {
+        if (iter->second->deadline_ != MonotonicTimePoint::zero_value) {
+          reschedule_deadline(iter->second, now);
+        }
+      }
+    }
+  }
+}
+
+void DataReaderImpl::reschedule_deadline(SubscriptionInstance_rch instance,
+                                         const MonotonicTimePoint& now)
+{
+  ACE_GUARD(ACE_Recursive_Thread_Mutex, guard, sample_lock_);
+
+  // So the datareader can call back into us.
+  if (instance->deadline_ != MonotonicTimePoint::zero_value) {
+
+    // Remove.
+    for (DeadlineQueue::iterator pos = deadline_queue_.lower_bound(instance->deadline_), limit = deadline_queue_.upper_bound(instance->deadline_); pos != limit; ++pos) {
+      if (pos->second == instance) {
+        deadline_queue_.erase(pos);
+        break;
+      }
+    }
+
+    instance->deadline_ = now + (deadline_period_ - (instance->deadline_ - now));
+
+    const bool schedule = deadline_queue_.empty();
+    deadline_queue_.insert(std::make_pair(instance->deadline_, instance));
+    if (schedule) {
+      deadline_task_->schedule(deadline_period_);
+    } else if (deadline_queue_.begin()->second == instance) {
+      // Moved to front.
+      deadline_task_->cancel();
+      deadline_task_->schedule(deadline_period_);
+    }
+  }
+}
+
+void DataReaderImpl::deadline_task(const MonotonicTimePoint& now)
+{
+  ThreadStatusManager::Event ev(TheServiceParticipant->get_thread_status_manager());
+
+  ACE_GUARD(ACE_Recursive_Thread_Mutex, guard, sample_lock_);
+  for (DeadlineQueue::iterator pos = deadline_queue_.begin(), limit = deadline_queue_.end(); pos != limit && pos->first <= now;) {
```

Comment:
For consistency, might want to use lower bound here?

Reply:
For `limit`?  I would prefer to not to do a separate lookup.  It doesn't make sense for `pos` because the front of the queue needs to be processed.

Reply:
Yeah, for limit. It's O(logn) timepoint comparisons to find the stopping point (via a single lower_bound lookup) rather than O(n) for comparing every iterator's timepoint against `now`. It's not a big deal, performance-wise, but above you're using both lower and upper bound to determine your start / stop positions so I thought it would make sense for consistency to do it here as well.

Reply:
I'm assuming that the actual number of elements processed is small, i.e., less than `log N`, so that it would still be more efficient to compare instead of doing the lookup.

Reply:
Gotcha. Hadn't considered that we're usually waiting on the first deadline, meaning we often / usually only need to consider the first element in the map.